### DISCUSSION TITLE
promtool responds with success even if the series input doesn't span to the eval_time window

### DISCUSSION THREAD

**Discussion by saimanoj90** [prometheus/prometheus#14504] [Q&A]
### What did you do?

alerts.yaml

groups:
- name: example
  rules:

  - alert: InstanceDown
    expr: up == 0
    for: 5m
    labels:
        severity: page
    annotations:
        summary: "Instance {{ $labels.instance }} down"
        description: "{{ $labels.instance }} of job {{ $labels.job }} has been down for more than 5 minutes."

---

test.yaml

rule_files:
    - alerts.yml

evaluation_interval: 1m

tests:
    # Test 1.
    - interval: 1m
      # Series data.
      input_series:
          - series: 'up{job="prometheus", instance="localhost:9090"}'
            values: '0 0 0 0 0 0'
      # Unit test for alerting rules.
      alert_rule_test:
          # Unit test 1.
          - eval_time: 10m
            alertname: InstanceDown
            exp_alerts:
                # Alert 1.
                - exp_labels:
                      severity: page
                      instance: localhost:9090
                      job: prometheus
                  exp_annotations:
                      summary: "Instance localhost:9090 down"
                      description: "localhost:9090 of job prometheus has been down for more than 5 minutes."

eval_time is 10m but the series input is only for 6m hence the the alert shouldn't be triggered but the alert is triggering. 

### What did you expect to see?

Unit Testing:  /data/test.yaml
  FAILED:
    alertname: InstanceDown, time: 10m,
        exp:[
            0:
              Labels:{alertname="InstanceDown", instance="localhost:9090", job="prometheus", severity="page"}
              Annotations:{description="localhost:9090 of job prometheus has been down for more than 5 minutes.", summary="Instance localhost:9090 down"}
            ],
        got:[]

### What did you see instead? Under which circumstances?

Unit Testing:  /data/test.yaml
  SUCCESS

### Prometheus version

```text
promtool, version 2.52.0 (branch: HEAD, revision: 879d80922a227c37df502e7315fad8ceb10a986d)
  build user:       root@1b4f4c206e41
  build date:       20240508-21:56:43
  go version:       go1.22.3
  platform:         linux/amd64
  tags:             netgo,builtinassets,stringlabels
```

**Comment by beorn7** [comment]
I think the behavior is correct.

Your test series has only samples in the first 6m, but that doesn't mean the series ends there. In the way you provided the test data, the implication is that there were just no scrapes happening anymore. In that case, the series stays "alive" until the lookback-delta has passed (by default 5m). So at the evaluation time specified, the series is still considered to have a value of 0, and your alert fires.

To mark the end of a series, so that you don't have to wait for the lookback-delta to pass, you can use staleness markers (keyword `stale` in the test data, see https://prometheus.io/docs/prometheus/latest/configuration/unit_testing_rules/#series for details).

Try `values: '0 0 0 0 0 0 stale'` in your example above.

I'll convert this to a discussion as it doesn't seem to be a bug in Prometheus.

**Comment by saimanoj90** [comment]
Is it possible to write unit tests for alerts which have more than one metric in the expression? for example

    - alert: NodeMemoryUsageWarnAlert
      expr: (avg_over_time(node_memory_MemAvailable_bytes[10m])/node_memory_MemTotal_bytes*100 < 10) and (avg_over_time(node_memory_MemAvailable_bytes[10m])/node_memory_MemTotal_bytes*100 > 5)
      for: 10m
